File: Stripe/STPPaymentIntentShippingDetailsParams.m
//
//  STPPaymentIntentShippingDetailsParams.m
//  Stripe
//
//  Created by Yuki Tokuhiro on 4/27/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

#import "STPPaymentIntentShippingDetailsParams.h"

#import "STPPaymentIntentShippingDetailsAddressParams.h"

@implementation STPPaymentIntentShippingDetailsParams

- (instancetype)initWithAddress:(STPPaymentIntentShippingDetailsAddressParams *)address name:(NSString *)name {
    self = [super init];
    if (self) {
        _address = address;
        _name = [name copy];
    }
    return self;
}

- (NSString *)description {
    NSArray *props = @[
                       // Object
                       [NSString stringWithFormat:@"%@: %p", NSStringFromClass([self class]), self],
                       
                       // Properties
                       [NSString stringWithFormat:@"address = %@", self.address],
                       [NSString stringWithFormat:@"name = %@", self.name],
                       [NSString stringWithFormat:@"carrier = %@", self.carrier],
                       [NSString stringWithFormat:@"phone = %@", self.phone],
                       [NSString stringWithFormat:@"trackingNumber = %@", self.trackingNumber],
                       ];
    
    return [NSString stringWithFormat:@"<%@>", [props componentsJoinedByString:@"; "]];
}

#pragma mark - NSCopying

- (instancetype)copyWithZone:(NSZone *)zone {
    __typeof(self) copy = [[[self class] allocWithZone:zone] init];

    copy.address = [self.address copy];
    copy.name = self.name;
    copy.carrier = self.carrier;
    copy.phone = self.phone;
    copy.trackingNumber = self.trackingNumber;

    return copy;
}

#pragma mark - STPFormEncodable

@synthesize additionalAPIParameters;

+ (nonnull NSDictionary *)propertyNamesToFormFieldNamesMapping {
    return @{
             NSStringFromSelector(@selector(address)): @"address",
             NSStringFromSelector(@selector(name)): @"name",
             NSStringFromSelector(@selector(carrier)): @"carrier",
             NSStringFromSelector(@selector(phone)): @"phone",
             NSStringFromSelector(@selector(trackingNumber)): @"tracking_number",
             };
}

+ (nullable NSString *)rootObjectName {
    return nil;
}

@end

File: Stripe/STPPaymentIntentShippingDetailsParams.swift
//
//  STPPaymentIntentShippingDetailsParams.swift
//  Stripe
//
//  Created by Yuki Tokuhiro on 4/27/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

import Foundation

/// Shipping information for a PaymentIntent
/// - seealso: https://stripe.com/docs/api/payment_intents/confirm#confirm_payment_intent-shipping
public class STPPaymentIntentShippingDetailsParams: NSObject {

    /// Shipping address.
    @objc public var address: STPPaymentIntentShippingDetailsAddressParams

    /// Recipient name.
    @objc public var name: String

    /// The delivery service that shipped a physical product, such as Fedex, UPS, USPS, etc.
    @objc public var carrier: String?

    /// Recipient phone (including extension).
    @objc public var phone: String?

    /// The tracking number for a physical product, obtained from the delivery service. If multiple tracking numbers were generated for this purchase, please separate them with commas.
    @objc public var trackingNumber: String?

    /// :nodoc:
    @objc public var additionalAPIParameters: [AnyHashable: Any] = [:]

    /// Initialize an `STPPaymentIntentShippingDetailsParams` with required properties.
    @objc
    public init(address: STPPaymentIntentShippingDetailsAddressParams, name: String) {
        self.address = address
        self.name = name
        super.init()

    }

    /// :nodoc:
    @objc public override var description: String {
        let props: [String] = [
            // Object
            String(
                format: "%@: %p", NSStringFromClass(STPPaymentIntentShippingDetailsParams.self),
                self),
            // Properties
            "address = \(address)",
            "name = \(name)",
            "carrier = \(String(describing: carrier))",
            "phone = \(String(describing: phone))",
            "trackingNumber = \(String(describing: trackingNumber))",
        ]

        return "<\(props.joined(separator: "; "))>"
    }

}

// MARK: - STPFormEncodable
extension STPPaymentIntentShippingDetailsParams: STPFormEncodable {

    @objc
    public class func propertyNamesToFormFieldNamesMapping() -> [String: String] {
        return [
            NSStringFromSelector(#selector(getter:address)): "address",
            NSStringFromSelector(#selector(getter:name)): "name",
            NSStringFromSelector(#selector(getter:carrier)): "carrier",
            NSStringFromSelector(#selector(getter:phone)): "phone",
            NSStringFromSelector(#selector(getter:trackingNumber)): "tracking_number",
        ]
    }

    @objc
    public class func rootObjectName() -> String? {
        return nil
    }

}

// MARK: - NSCopying
extension STPPaymentIntentShippingDetailsParams: NSCopying {
    /// :nodoc:
    @objc
    public func copy(with zone: NSZone? = nil) -> Any {
        let copy = STPPaymentIntentShippingDetailsParams(address: address, name: name)

        copy.carrier = carrier
        copy.phone = phone
        copy.trackingNumber = trackingNumber
        copy.additionalAPIParameters = additionalAPIParameters

        return copy
    }

}
